### ISSUE TITLE
use_self: breaks code with type aliases

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9917]
### Summary

.

### Lint Name

use_self

### Reproducer

I tried this code:

```rust
#![allow(non_snake_case)]
struct Struct { }
type Alias = Struct;

impl Struct {
    fn ref_Alias(self: &Alias, f: &u32) -> &u32 {
        f
    }
}

fn main() { }
```

I saw this happen:
`cargo clippy --fix -- -Aclippy::all -Wclippy::use_self`
```
    Checking clpy v0.1.0 (/tmp/clpy)
warning: failed to automatically apply fixes suggested by rustc to crate `clpy`

after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error: lifetime may not live long enough
 --> src/main.rs:7:9
  |
6 |     fn ref_Alias(self: &Self, f: &u32) -> &u32 {
  |                        -         - let's call the lifetime of this reference `'1`
  |                        |
  |                        let's call the lifetime of this reference `'2`
7 |         f
  |         ^ associated function was supposed to return data with lifetime `'2` but it is returning data with lifetime `'1`
  |
help: consider introducing a named lifetime parameter and update trait if needed
  |
6 |     fn ref_Alias<'a>(self: &'a Self, f: &'a u32) -> &u32 {
  |                 ++++        ++           ++

error: aborting due to previous error

Original diagnostics will follow.

error: lifetime may not live long enough
 --> src/main.rs:7:9
  |
6 |     fn ref_Alias(self: &Self, f: &u32) -> &u32 {
  |                        -         - let's call the lifetime of this reference `'1`
  |                        |
  |                        let's call the lifetime of this reference `'2`
7 |         f
  |         ^ associated function was supposed to return data with lifetime `'2` but it is returning data with lifetime `'1`
  |
help: consider introducing a named lifetime parameter and update trait if needed
  |
6 |     fn ref_Alias<'a>(self: &'a Self, f: &'a u32) -> &u32 {
  |                 ++++        ++           ++

error: could not compile `clpy` due to previous error
warning: build failed, waiting for other jobs to finish...
warning: struct `Struct` is never constructed
 --> src/main.rs:2:8
  |
2 | struct Struct { }
  |        ^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: type alias `Alias` is never used
 --> src/main.rs:3:6
  |
3 | type Alias = Struct;
  |      ^^^^^

warning: associated function `ref_Alias` is never used
 --> src/main.rs:6:8
  |
6 |     fn ref_Alias(self: &Alias, f: &u32) -> &u32 {
  |        ^^^^^^^^^

warning: unnecessary structure name repetition
 --> src/main.rs:6:25
  |
6 |     fn ref_Alias(self: &Alias, f: &u32) -> &u32 {
  |                         ^^^^^ help: use the applicable keyword: `Self`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#use_self
  = note: requested on the command line with `-W clippy::use-self`

warning: `clpy` (bin "clpy") generated 4 warnings (run `cargo fix --bin "clpy"` to apply 1 suggestion)
```

### Version

```text
rustc 1.67.0-nightly (c5d82ed7a 2022-11-19)
binary: rustc
commit-hash: c5d82ed7a4ad94a538bb87e5016e7d5ce0bd434b
commit-date: 2022-11-19
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

**Comment by xtqqczze** [comment]
Still an issue with Rust 1.92: https://godbolt.org/z/sn7z4M8oj
